flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/SparseVector.java

```diff
@@ -0,0 +1,725 @@
+				r.add(this.indices[i], this.values[i]);
+			}
+			return r;
+		} else {
+			return SparseVector.apply(this, (SparseVector) vec, ((a, b) -> a + b));
+		}
+	}
+
+	@Override
+	public Vector minus(Vector vec) {
+		if (this.size() != vec.size()) {
+			throw new IllegalArgumentException("The size of the two vectors are different.");
+		}
+
+		if (vec instanceof DenseVector) {
+			DenseVector r = ((DenseVector) vec).scale(-1.0);
+			for (int i = 0; i < this.indices.length; i++) {
+				r.add(this.indices[i], this.values[i]);
+			}
+			return r;
+		} else {
+			return SparseVector.apply(this, (SparseVector) vec, ((a, b) -> a - b));
+		}
+	}
+
+	@Override
+	public SparseVector scale(double d) {
+		SparseVector r = new SparseVector(this.n, this.indices, this.values);
+		for (int i = 0; i < this.values.length; i++) {
+			r.values[i] *= d;
+		}
+		return r;
+	}
+
+	@Override
+	public void scaleEqual(double d) {
+		for (int i = 0; i < this.values.length; i++) {
+			this.values[i] *= d;
+		}
+	}
+
+	/**
+	 * Remove all zero values away from this vector.
+	 */
+	public void removeZeroValues() {
+		if (this.values.length != 0) {
+			List<Integer> idxs = new ArrayList<>();
+			for (int i = 0; i < values.length; i++) {
+				if (0 != values[i]) {
+					idxs.add(i);
+				}
+			}
+			int[] newIndices = new int[idxs.size()];
+			double[] newValues = new double[newIndices.length];
+			for (int i = 0; i < newIndices.length; i++) {
+				newIndices[i] = indices[idxs.get(i)];
+				newValues[i] = values[idxs.get(i)];
+			}
+			this.indices = newIndices;
+			this.values = newValues;
+		}
+	}
+
+	private double dot(SparseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("the size of the two vectors are different");
+		}
+
+		double d = 0;
+		int p0 = 0;
+		int p1 = 0;
+		while (p0 < this.values.length && p1 < other.values.length) {
+			if (this.indices[p0] == other.indices[p1]) {
+				d += this.values[p0] * other.values[p1];
+				p0++;
+				p1++;
+			} else if (this.indices[p0] < other.indices[p1]) {
+				p0++;
+			} else {
+				p1++;
+			}
+		}
+		return d;
+	}
+
+	private double dot(DenseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException(
+				"The size of the two vectors are different: " + this.size() + " vs " + other.size());
+		}
+		double s = 0.;
+		for (int i = 0; i < this.indices.length; i++) {
+			s += this.values[i] * other.get(this.indices[i]);
+		}
+		return s;
+	}
+
+	@Override
+	public double dot(Vector other) {
+		if (other instanceof DenseVector) {
+			return dot((DenseVector) other);
+		} else {
+			return dot((SparseVector) other);
+		}
+	}
+
+	/**
+	 * Compute the outer product with itself.
+	 *
+	 * @return The outer product matrix.
+	 */
+	public DenseMatrix outer() {
+		return this.outer(this);
+	}
+
+	/**
+	 * Compute the outer product with another vector.
+	 *
+	 * @return The outer product matrix.
+	 */
+	public DenseMatrix outer(SparseVector other) {
```

Comment:
Same here. `outer` should be a generic API across sparse/dense vectors

Reply:
Currently there are no situations where the outer product between a dense vector and a sparse vector is needed. Perhaps we may postpone until the requirement shows.